File: remotedb/src/main/kotlin/com/novoda/remotedb/RemoteDatabaseProvider.java
package com.novoda.remotedb;

import android.content.Context;

import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseOptions;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

public enum RemoteDatabaseProvider {

    INSTANCE;

    private FirebaseApp firebaseApp;

    public RemoteDatabase obtainNode(Context context, RemoteDatabaseStructure remoteDatabaseStructure) {
        initialiseFirebaseAppIfNecessary(context);
        FirebaseDatabase instance = FirebaseDatabase.getInstance(firebaseApp);
        DatabaseReference reference = instance.getReference(remoteDatabaseStructure.getFirstChildId());
        for (String childId : remoteDatabaseStructure.getChildIds()) {
            reference = reference.child(childId);
        }
        return new RemoteDatabase(reference);
    }

    private void initialiseFirebaseAppIfNecessary(Context context) {
        if (firebaseApp == null) {
            FirebaseOptions firebaseOptions = new FirebaseOptions.Builder()
                    .setApplicationId("com.novoda.materialised") // build() throws an exception if this isn't set
                    .setDatabaseUrl("https://hacker-news.firebaseio.com")
                    .build();
            firebaseApp = FirebaseApp.initializeApp(context.getApplicationContext(), firebaseOptions, "Materialised");
        }
    }

}

File: remotedb/src/main/kotlin/com/novoda/remotedb/RemoteDatabaseProvider.kt
package com.novoda.remotedb

import android.content.Context
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.database.FirebaseDatabase

class RemoteDatabaseProvider private constructor(private val firebaseApp: FirebaseApp) {

    fun node(structure: Structure): RemoteDatabase {
        val instance = FirebaseDatabase.getInstance(firebaseApp)
        var reference = instance.getReference(structure.firstChildId)
        for (childId in structure.childIds) {
            reference = reference.child(childId)
        }
        return RemoteDatabase(reference)
    }

    companion object {

        private var lastUsedMetadata: Metadata? = null // wanted to use lateinit here too, but isInitialized doesn't work for some reason
        private lateinit var lastReturnedProvider: RemoteDatabaseProvider

        fun obtain(context: Context, metadata: Metadata): RemoteDatabaseProvider =
                when {
                    lastUsedMetadata != null && metadata == lastUsedMetadata -> lastReturnedProvider
                    else -> createNewProvider(context, metadata)
                }

        private fun createNewProvider(context: Context, metadata: Metadata): RemoteDatabaseProvider {
            val firebaseApp = FirebaseApp.initializeApp(
                    context.applicationContext,
                    FirebaseOptions.Builder().containing(metadata),
                    metadata.applicationName
            )
            lastUsedMetadata = metadata
            lastReturnedProvider = RemoteDatabaseProvider(firebaseApp)
            return lastReturnedProvider
        }

        private fun FirebaseOptions.Builder.containing(metadata: Metadata): FirebaseOptions = apply {
            setApplicationId(metadata.applicationId)
            setDatabaseUrl(metadata.databaseUrl)
        }.build()
    }
}
